### DISCUSSION TITLE
RDS Enhancements Proposal

### DISCUSSION THREAD

**Discussion by maciejwalkowiak** [awspring/spring-cloud-aws#82] [Ideas]
From spring-cloud-aws created by [maciejwalkowiak](https://github.com/maciejwalkowiak): spring-cloud/spring-cloud-aws#610  

Few ideas on how RDS integration could be improved:

RDS integration was created before Spring Boot. For user experienced with Spring Boot datasource configuration, using Spring Cloud AWS RDS integration may not be as seamless as it could be. Things that Spring Cloud AWS should provide:
- connection pool customization through autoconfiguration (like in Spring Boot) - we should not enforce Tomcat pool
- datasource configuration through `DatasourceProperties`

New features:
- seamless integration with [aws-secretsmanager-jdbc](https://github.com/aws/aws-secretsmanager-jdbc) - presence of the library on the classpath could trigger autoconfiguration that sets JDBC driver
- Aurora support

Configuring RDS integration could look following:
```
spring.datasource.username=<user>
spring.datasource.password=<pass>
... # other Spring Boot datasource properties, but no JDBC url needed
cloud.aws.rds.instance-id=<instance-id> # or cloud.aws.rds.cluster-id in case of Aurora
cloud.aws.rds.db-name=postgres
cloud.aws.rds.read-replica=true
cloud.aws.rds.jdbc-url-query= # to set extra parametrs on generated JDBC url
```

I am not sure if we should keep supporting multiple database instance or also drop it together with Spring Boot alignment (once Spring Boot has opinionated view on configuring multiple databases we could align with it too).

**Comment by maciejwalkowiak** [comment]
This ⬆️   is up for discussion so if you have an idea or there is something related to RDS integration you believe we should include - your comments are very welcome!

**Comment by maciejwalkowiak** [comment]
It would be a good idea to mention about the configuration support towards multiple datasources against a singe RDS instance. Here the usecase is of an application which is connecting to multiple databases within a single instance of RDS. 

Would be good to align as Spring Boot currently does https://docs.spring.io/spring-boot/docs/current/reference/html/howto.html#howto-two-datasources

**Comment by eddumelendez** [comment]
We already talked about this, but leaving my comments just for the record. IMO this module can be removed after the evaluation of [RDS Proxy](https://aws.amazon.com/rds/proxy/)

Why?
* Currently, [aws-secretsmanager-jdbc](https://github.com/aws/aws-secretsmanager-jdbc) only supports aws sdk v1 and jdbc
* AWS Secrets Manager can be created with a different KMS and this is not managed by the driver AFAIK.
* RDS Proxy can be the solution for previous `spring-cloud-aws` versions and new ones related to `jdbc` and `r2dbc` integration and also reduction time for database failover and of course the integration with AWS Secrets Manager.

**Reply by maciejwalkowiak** [reply]
- `aws-secretsmanager-jdbc` supporting only aws sdk v1 is a non-issue - v1 and v2 can live side by side in single project
- RDS Proxy looks good, it solves the security and failover problems really well. It has certain [limitations](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/rds-proxy.html#rds-proxy-setup) though - not sure how severe they are. Can we support read replicas with RDS Proxy? To keep the functionality of `@Transactional(readOnly=true)` going to read-replica? I think this is where Spring Cloud AWS really shines with RDS support. 

Considering [RDS Proxy have now read-only endpoints](https://aws.amazon.com/about-aws/whats-new/2021/03/amazon-rds-proxy-adds-read-only-endpoints-for-amazon-aurora-replicas/), maybe we can drop failover from Spring Cloud AWS and keep read replica support, including support through RDS Proxy?

**Comment by v-sag** [comment]
Hi,

Any update on integration with `aws-secretsmanager-jdbc`? I would like to use the `spring-cloud-aws-jdbc` feature for auto-detecting read replicas, but at the same time to use rotated password; I couldn't find anything on the latest documentation that I could do that.
